src/rules/optimize-image/optimize-image.ts

```diff
@@ -0,0 +1,124 @@
+
+/*
+ * ------------------------------------------------------------------------------
+ * Public
+ * ------------------------------------------------------------------------------
+ */
+
+const rule: IRuleBuilder = {
+    create(context: RuleContext): IRule {
+        /** Error processing the request. */
+        let failed: boolean = false;
+        /** The promise that represents the analyzing result. */
+        let analyzePromise: Promise<Result>;
+
+        const username: string = process.env.SERVICE_USERNAME; // eslint-disable-line no-process-env
+        const endpoint: string = process.env.API_ENDPOINT; // eslint-disable-line no-process-env
+        const requestAsync = require('request-promise');
+
+        const notifyError = async (resource: string, error: any) => {
+            debug(`Error getting analyzing result for ${resource}.`, error);
+            await context.report(resource, null, `Couldn't get results for ${resource}. Error: ${error.message}`);
+        };
+
+        const validateTargetFetchEnd = (targetFetchEnd: ITargetFetchEnd) => {
+
+            const { resource }: { resource: string } = targetFetchEnd;
+
+            debug(`Validating rule.`);
+
+            if (!username) {
+                notifyError(resource, new Error('No username is provided for authentication.'));
+                failed = true;
+
+                return;
+            }
+
+            const options = {
+                auth: { user: username },
+                body: { url: resource },
+                json: true,
+                method: 'POST',
+                uri: endpoint
+            };
+
+            analyzePromise = requestAsync(options);
+            analyzePromise.catch(async (error) => {
+                failed = true;
+
+                await notifyError(resource, error);
+            });
+        };
+
+        const valiateScanEnd = async (fetchEnd: IScanEnd) => {
+            const { resource }: { resource: string } = fetchEnd;
+            const reportPromises: Array<Promise<void>> = [];
+
+            if (!analyzePromise || failed) {
+                return;
+            }
+
+            debug(`Waiting for the optimizing result of ${resource}.`);
+
+            const result: Result = await analyzePromise;
+
+            debug(`Received the optimizing result of ${resource}.`);
+
+            const { files }: { files: Array<File> } = result;
+            const unoptimized: Array<File> = files.filter((file) => {
+                if (!file) {
+                    return false;
+                }
+
+                const { bytes }: { bytes: Bytes } = file;
+
+                return bytes.output < bytes.input;
+            });
+
+            if (!unoptimized.length) {
+                debug(`All images are optimized.`);
+
+                return;
+            }
+
+            unoptimized.forEach((file) => {
+                const { bytes }: { bytes: Bytes } = file;
+                const sizeDiff: number = bytes.savings / 1000;
```

Comment:
`const sizeDiff: number = bytes.savings / 1024;`

A kB is 1024B.

Reply:
@sarvaje Done!